### PR TITLE
(getgov-bob) Ticket #1530: Warning messages

### LINKED ISSUE #1530: Add warning messages for changing statuses in admin portal
### Issue description

As an analyst
I want to receive warning messages when I put a domain on hold, mark it as ineligible, or delete a domain 
so that I do not accidently make a major change unintentionally

### Acceptance criteria

- [ ] implement design as defined in #1473
- [ ] user receive warning messages when a domain moves to the following status:
1. [ ] placed on hold
2. [ ] marked as ineligible
3. [ ] marked for delete
- [ ] Link design approach to dev ticket for implementation

### Additional context

[Slack thread for context](https://gsa-tts.slack.com/archives/C05BDEA3C11/p1702312238745639)

### Links to other issues

Block by: #1473 (design ticket)

### PR DESCRIPTION
**Pull request by zandercymatics**
## Ticket

Resolves #1530

## Changes

- `admin.py`
  - Added a custom change form template for domain application (for the delete modal)
  - Modified delete_view for DomainAdmin to include a custom delete confirmation page
- `get-gov-admin.js`
  - Modified the `createPhantomModalFormButtons` function to make it more specific
  - Added a `displayModalOnDropdownClick` function that displays a modal on status change
- `_admin.scss`
  - Added styling to li elements inside modals to match <p> tags.
- Added `domain_application_change_form.html`
- Modified `domain_change_form.html`
- Added `domain_delete_confirmation.html`
- Added tests in `test_admin.py` 

## Context for reviewers
On Django Admin, this ticket adds confirmation modals in three places:
1. When the "Place hold" button is pressed (Found with a Domain in the state "ready")
2. When the "Remove from registry" button is pressed
3.  When the status on a DomainRequest is changed to "ineligible" (before save)

A custom delete confirmation message is displayed in these two places:
1. When the "delete" button is pressed for a domain object
2. When the "delete selected domains" button is pressed on the tables

## Setup
1. Navigate to Django Admin
2. Navigate to Domain Request and select a record
3. Change the status of a domain to "ineligible". A modal should display. When clicking cancel, the status should revert.
4. Navigate to Domain and select a record in the state "Ready"
5. The "Place hold" and "Remove from registry" buttons (pressed in that order) should both display modals
6. Click the "delete" button on the bottom of the page. The delete confirmation page that you get redirected to should have custom content 
7. Navigate back to the Domain table view. Delete a record from there. Like before, a custom delete confirmation page should be presented.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

**Comment by erinysong** [comment]
Deployed this to my sandbox if anyone else can benefit from using it

**Comment by erinysong** [comment]
UI question: Do we want the modal buttons be bullet points? This may have been already discussed and I know Django admin has limited UI capabilities, but I was wondering if it'd be possible to remove them. Minor nit since this will be analyst facing and acknowledge that functionality takes precedence
image

**Comment by erinysong** [comment]
This is more of a design question but what should happen if I visit a domain application record, make the status Ineligible to prompt the warning modal, and click cancel? Should the status still remain Ineligible or should it switch back to the previous status? Currently when I click cancel/away from the warning modal, the status still remains ineligible. Since this is a warning modal and for an important-but-nonbreaking change, I'm wondering if there's a way to inform the user they're making a domain application Ineligible but not require them to acknowledge that they're doing so.
image
image

**Comment by erinysong** [comment]
May be misunderstanding what Place Hold does, but when I create a add domain from admin to test, I have the option to Place Hold and from the Place Hold modal, I can directly create a new domain that does not have a Hold status. I was wondering if Place Hold should be accessible from the Add a New Domain page and if so, should the modal do this?

1. Create a new domain hold-from-add-domain.gov
image
2. Click on the Place Hold button, prompting the Place Hold warning modal. The status is blank for Domain or New status.  
image
3. Clicking the confirmation button on the Place Hold modal adds the domain. 
image
4. The status of the domain is Unknown despite placing a hold on the domain.
image

**Comment by erinysong** [comment]
Thank you for the heads up about the static assets - that successfully removed the bullet points and got the right behavior on the cancel button! I also ran the commands you gave and updated them on my sandbox. The domain application status now reverts back to its initial status if you click cancel on the modal, but it still seems to switch to Ineligible if you close the modal by clicking the close button or clicking outside of the modal. Should we add the same logic we have to the cancel button to the close modal button and the event the user clicks away from the button?
Including the flow I used for visual reference

1. Domain application is in review. Changed the status to Ineligible
image
2. Click on the close button on the top right corner of the modal or anywhere outside of the modal. The status change to Ineligible goes through as opposed to the cancel button, which reverts the domain application to its initial status
image
image

File: src/registrar/templates/django/admin/domain_application_change_form.html
Comment (erinysong) [inline src/registrar/templates/django/admin/domain_application_change_form.html:83]: Should we also connect the same logic used for "_cancel_application_ineligible" to this button + the event the user clicks outside of the modal?

**Review by erinysong — CHANGES_REQUESTED** [review]
The modals look great! Wrote down on a couple edge cases but let me know if any of these are a result of static assets that aren't properly on my sandbox. I'm currently unable to access the on hold modal flow because all my domains have status Unknown which I'll try to fix within my sandbox, but posting here in case anyone is able to test the On Hold flow before I'm able to fix my data.

1. Replicating the modal cancel button's behavior when you also close the close button and/or click outside of the modal
2. Putting a domain on hold when you add domain.

**Comment by erinysong** [comment]
Successfully accessed Place Hold modal on domain in Ready state. User can also remove hold on a domain without being prompted by a warning modal.
image

**Review by erinysong — APPROVED** [review]
Confirmed that all warning modals and pages display successfully! Looks good to go after merging with main 🎉
Have a non-blocking comment on if we'd want to also have a warning modal when a user removes a hold on a domain, but that is outside the scope of this ticket since it'd depend on design and content decisions.

**Review by Katherine-Osos — APPROVED** [review]

**Pull request merged**
